cudax/include/cuda/experimental/__container/async_buffer.cuh

```diff
@@ -0,0 +1,889 @@
+  _CCCL_TEMPLATE(class... _OtherProperties)
+  _CCCL_REQUIRES(__properties_match<_OtherProperties...>)
+  _CCCL_HIDE_FROM_ABI explicit async_buffer(async_buffer<_Tp, _OtherProperties...>&& __other) noexcept
+      : __buf_(_CUDA_VSTD::move(__other.__buf_))
+      , __size_(_CUDA_VSTD::exchange(__other.__size_, 0))
+      , __policy_(_CUDA_VSTD::exchange(__other.__policy_, __policy_t::invalid_execution_policy))
+  {}
+
+  //! @brief Constructs an empty async_buffer using an environment
+  //! @param __env The environment providing the needed information
+  //! @note No memory is allocated.
+  _CCCL_HIDE_FROM_ABI async_buffer(const __env_t& __env)
+      : async_buffer(__env, 0, ::cuda::experimental::uninit)
+  {}
+
+  //! @brief Constructs a async_buffer of size \p __size using a memory resource and value-initializes \p __size
+  //! elements
+  //! @param __mr The memory resource to allocate the async_buffer with.
+  //! @param __size The size of the async_buffer. Defaults to zero
+  //! @note If `__size == 0` then no memory is allocated.
+  _CCCL_HIDE_FROM_ABI explicit async_buffer(const __env_t& __env, const size_type __size)
+      : async_buffer(__env, __size, ::cuda::experimental::uninit)
+  {
+    if (__size != 0)
+    {
+      this->__value_initialize_n(__unwrapped_begin(), __size);
+    }
+  }
+
+  //! @brief Constructs a async_buffer of size \p __size using a memory resource and copy-constructs \p __size elements
+  //! from \p __value
+  //! @param __mr The memory resource to allocate the async_buffer with.
+  //! @param __size The size of the async_buffer.
+  //! @param __value The value all elements are copied from.
+  //! @note If `__size == 0` then no memory is allocated.
+  _CCCL_HIDE_FROM_ABI explicit async_buffer(const __env_t& __env, const size_type __size, const _Tp& __value)
+      : async_buffer(__env, __size, ::cuda::experimental::uninit)
+  {
+    if (__size != 0)
+    {
+      this->__fill_n(__unwrapped_begin(), __size, __value);
+    }
+  }
+
+  //! @brief Constructs a async_buffer of size \p __size using a memory and leaves all elements uninitialized
+  //! @param __mr The memory resource to allocate the async_buffer with.
+  //! @param __size The size of the async_buffer.
+  //! @warning This constructor does *NOT* initialize any elements. It is the user's responsibility to ensure that the
+  //! elements within `[vec.begin(), vec.end())` are properly initialized, e.g with `cuda::std::uninitialized_copy`.
+  //! At the destruction of the \c async_buffer all elements in the range `[vec.begin(), vec.end())` will be destroyed.
+  _CCCL_HIDE_FROM_ABI explicit async_buffer(const __env_t& __env, const size_type __size, ::cuda::experimental::uninit_t)
+      : __buf_(
+          __env.query(::cuda::experimental::get_memory_resource), __env.query(::cuda::experimental::get_stream), __size)
+      , __size_(__size)
+      , __policy_(__env.query(::cuda::experimental::execution::get_execution_policy))
+  {}
```

Comment:
Q: Should we statically assert that value type is trivially constructible?

Reply:
The Type must be trivially copy constructible which is probably sufficient here